USB replication: Add grub entries, improve set image handling

chaitu236 opened the pull request:
This patch set has rudimentary implementation for Get/Set Image Grub menu options in provisioning media.

New grub menu entries:

![image](https://github.com/ni/meta-nilrt/assets/8194523/86a804bd-f626-4323-93ae-5be3d7c50e91)

`Get Image` entry output:

![image](https://github.com/ni/meta-nilrt/assets/8194523/72e21c73-a1fe-400c-a76c-5a47d1c43150)

`Set Image` entry output:

![image](https://github.com/ni/meta-nilrt/assets/8194523/1a769fd1-9724-41fb-8610-7779fce98420)

Patch set has following changes

- Add grub menu entries for Get/Set Image
- Use `bootdelay` and `consoleoutputenable` from image but initialize other grubenv entries to defaults.
- Add missing dependencies of ni-netcfgutil.
- Handle netcfg operations in setall.
- Fix populateconfig so it works in restoremode.

WI: [AB#2372259](https://dev.azure.com/ni/DevCentral/_workitems/edit/2372259)

### Testing
- [x] Tested "Get Image" grub entry on cRIO-9030 and a VM
- [x] Tested "Set Image" grub entry on cRIO-9030 with image from the same target and image from a VM
- [x] hostname, ssh keys are handled correctly i.e., they're not replicated

### Known issues
- Network configuration like static ip address setting are also being replicated which is probably not what we want. Will address this in another PR.
- Error handling (e.g., when media is read-only, etc), ability to provide image name, ability to pick a certain image will be addressed in another PR.

chaitu236 committed:
nisystemreplication: Initialize grubenv

Other than bootdelay and consoleoutputenable, we don't want any grubenv
variables to be carried over from another target.

So keep bootdelay and consoleoutputenable from the image and initialize
others with default values.

Also added some cleanup code from ni_provisioning.safemode.

Signed-off-by: Chaitanya Vadrevu <chaitanya.vadrevu@ni.com>

chaitu236 committed:
grub_old.cfg: Add grub menu options for usb replication

Implement get/set grub menu options for usb replication.

Increase grub timeout to 3s as 1s is too short to be able to select menu
options.

Signed-off-by: Chaitanya Vadrevu <chaitanya.vadrevu@ni.com>

chaitu236 committed:
ni-netcfgutil: Add required dependencies to RDEPENDS

Added following packages which are required dependencies

- ni-utils: Provides /etc/natinst/networking/functions.common
- initscripts: Provides /etc/init.d/hostname.sh
- initscripts-nilrt: Provides /etc/init.d/populateconfig

Signed-off-by: Chaitanya Vadrevu <chaitanya.vadrevu@ni.com>

chaitu236 committed:
nilrt-recovery-media: Install nirtcfg

nirtcfg is a dependency for ni-netcfgutil but since this is built
outside of OE, we need to add it here.

Signed-off-by: Chaitanya Vadrevu <chaitanya.vadrevu@ni.com>

chaitu236 committed:
nisystemimage: Handle netcfg operations in setall

- Handle netcfg operations in setall
- Refactor init_fs_restoremode for readability
- Add cleanup_fs_restoremode to correctly handle cleanup in restoremode

Signed-off-by: Chaitanya Vadrevu <chaitanya.vadrevu@ni.com>

chaitu236 committed:
populateconfig: Check for directory presence before creating links

populateconfig is going to be used in restoremode for usb replication
feature.

Since /var/local/natinst doesn't exist in restormode, check for it
before creating /var/local/natinst/certstore symlink.

Signed-off-by: Chaitanya Vadrevu <chaitanya.vadrevu@ni.com>

amstewart commented on recipes-ni/ni-systemreplication/files/nisystemreplication:
```diff
@@ -12,12 +12,34 @@ image_get () {
+	enable_automount
+	print_done
+
+	trap - ERR
+	exec 1>&3
+	exec 2>&4
```
What are these for?

chaitu236 replied:
Copied these from the regular provisioning codeflow https://github.com/ni/meta-nilrt/blob/nilrt/master/kirkstone/recipes-core/initrdscripts/files/ni_provisioning.safemode#L20.

I believe they're undoing the [custom error handler](https://github.com/ni/meta-nilrt/blob/nilrt/master/kirkstone/recipes-core/initrdscripts/files/ni_provisioning.common#L460) and IO redirection done earlier in [early_setup](https://github.com/ni/meta-nilrt/blob/nilrt/master/kirkstone/recipes-core/initrdscripts/files/ni_provisioning.common#L479).

amstewart commented on recipes-ni/ni-systemreplication/files/nisystemreplication:
```diff
@@ -12,12 +12,34 @@ image_get () {
 image_set () {
 	source /ni_provisioning.common
 	early_setup
+
 	source /ni_provisioning.safemode.common
+	install_safemode
+
 	nisystemimage setall -x tgz -f getall.tgz -p reset -s reset
 
-	# save rootfs UUID for grub to reference
-	ROOTUUID=`lsblk ${TARGET_DISK}${PART_SEPARATOR}4 -n -o PARTUUID`
-	echo set rootuuid=$ROOTUUID >> $GRUB_MOUNTPOINT/grubvar_readonly
+	# Retain some grubenv settings
+	if grep --quiet "^consoleoutenable=" $BOOTFS_MOUNTPOINT/grub/grubenv; then
+		grubenv_consoleoutenable=`grub-editenv $BOOTFS_MOUNTPOINT/grub/grubenv list |grep "^consoleoutenable=" | cut -f 2 -d '='`
+	fi
```
It isn't clear to me were these variables are getting used. If they're being used by the `install_grubenv` func below, can they just be calculated there?

chaitu236 replied:
They're used in `install_grubenv` [here](https://github.com/ni/meta-nilrt/blob/nilrt/master/kirkstone/recipes-core/initrdscripts/files/ni_provisioning.safemode.common#L304) and [here](https://github.com/ni/meta-nilrt/blob/nilrt/master/kirkstone/recipes-core/initrdscripts/files/ni_provisioning.safemode.common#L322).

They were originally added to be set from [early_setup](https://github.com/ni/meta-nilrt/blob/nilrt/master/kirkstone/recipes-core/initrdscripts/files/ni_provisioning.common#L502), but we can't determine the values for these until the `getall.tgz` archive is unpacked.. so we have to do this after `nisystemimage setall`.

> can they just be calculated there

`install_grubenv` needs to be able to work as before for regular provisioning.. can't think of a good way to move it into that script.

amstewart commented on recipes-ni/ni-systemreplication/files/nisystemreplication:
```diff
@@ -12,12 +12,34 @@ image_get () {
+		grubenv_consoleoutenable=`grub-editenv $BOOTFS_MOUNTPOINT/grub/grubenv list |grep "^consoleoutenable=" | cut -f 2 -d '='`
+	fi
+	if grep --quiet "^bootdelay=" $BOOTFS_MOUNTPOINT/grub/grubenv; then
+		grubenv_bootdelay=`grub-editenv $BOOTFS_MOUNTPOINT/grub/grubenv list |grep "^bootdelay=" | cut -f 2 -d '='`
```
NIT: Prefer `$(...)` to backtick styling.

chaitu236 replied:
Done

amstewart commented on recipes-ni/ni-systemreplication/files/nisystemreplication:
```diff
@@ -5,18 +5,28 @@ set -e
-	nisystemimage getall -x tgz -f getall.tgz
+	mount -o remount,rw $NIRECOVERY_MOUNTPOINT
+	mkdir -p $NIRECOVERY_MOUNTPOINT/replication_images
+	nisystemimage getall -x tgz -f $NIRECOVERY_MOUNTPOINT/replication_images/getall.tgz
```
MINOR: Can we figure out a more meaningful name for the archive than `getall.tgz`? Maybe: `latest.tgz`?

chaitu236 replied:
`latest.tgz` may also cause confusion about what kind of image that is - the one containing just runmode and config (i.e., get operation) or the one containing grub, safemode, runmode, config (i.e., getall).

How about `systemimage.tgz` or `latest_systemimage.tgz`?

amstewart replied:
I'd be fine with either. I suppose the former makes more sense if this workflow isn't supposed to support having multiple images on the tool at once.

chaitu236 replied:
Done. It's just one image for now but we'll expand it to multiple.

amstewart commented on scripts/lib/wic/canned-wks/grub_old.cfg:
```diff
@@ -1,6 +1,6 @@
 serial --unit=0 --speed=115200 --word=8 --parity=no --stop=1
 default='Default Provisioning'
-timeout=1
+timeout=3
```
What's the reason for extending the timeout? Just to make it easier to select the new options?

chaitu236 replied:
Yes, 1s was too short to be able to select new options.

amstewart replied:
OK. Could you just add that reasoning to the commit message. It's the kind of logic that someone is going to have to check in the git logs in the future.

chaitu236 replied:
Done

amstewart commented on recipes-ni/ni-systemimage/files/nisystemimage:
```diff
@@ -634,7 +641,32 @@ init_fd () {
 
+init_fs_restoremode () {
+	# If the util did not previously exit cleanly, it's possible that the
+	# working mount points didn't get unmount them. To avoid double-mounts,
```
Could you fixup the grammar in this comment you copied.

chaitu236 replied:
Done

amstewart commented on recipes-ni/ni-systemimage/files/nisystemimage:
```diff
@@ -634,7 +641,32 @@ init_fd () {
+
 init_fs () {
+	if in_restoremode; then
+		# Tools such as ninetcfg and nirtcfg same mountpoints as safemode and runmode
```
Grammar.

chaitu236 replied:
Done

amstewart reviewed (changes requested):
Since you have the context now, can we figure out a better name for the `grub_old.cfg` OE file?

chaitu236 committed:
grub.cfg: Replace grub.cfg with grub_old.cfg

grub_old.cfg is used for safemode provisioning and grub.cfg was for rauc
which we no longer use.

So replace grub.cfg with grub_old.cfg.

Signed-off-by: Chaitanya Vadrevu <chaitanya.vadrevu@ni.com>

chaitu236 commented:
> Since you have the context now, can we figure out a better name for the `grub_old.cfg` OE file?

Moved `grub_old.cfg` to `grub.cfg` as discussed.
